search/services/index.py

```diff
@@ -0,0 +1,785 @@
+
+from search.context import get_application_config, get_application_global
+from search import logging
+from search.domain import Document, DocumentSet, Query, DateRange, \
+    Classification, AdvancedQuery, SimpleQuery, AuthorQuery, Author, asdict
+
+logger = logging.getLogger(__name__)
+
+# TODO: make this configurable.
+MAX_RESULTS = 10_000
+"""This is the maximum result offset for pagination."""
+
+# We'll compile this ahead of time, since it gets called quite a lot.
+STRING_LITERAL = re.compile(r"(['\"][^'\"]*['\"])")
+"""Pattern for string literals (quoted) in search queries."""
+
+
+class MappingError(ValueError):
+    """There was a problem with the search document mapping."""
+
+
+class IndexConnectionError(IOError):
+    """There was a problem connecting to the search index."""
+
+
+class IndexingError(IOError):
+    """There was a problem adding a document to the index."""
+
+
+class QueryError(ValueError):
+    """
+    Elasticsearch could not handle the query.
+
+    This is likely due either to a programming error that resulted in a bad
+    index, or to a mal-formed query.
+    """
+
+
+class DocumentNotFound(RuntimeError):
+    """Could not find a requested document in the search index."""
+
+
+class OutsideAllowedRange(RuntimeError):
+    """A page outside of the allowed range has been requested."""
+
+
+def _wildcardEscape(querystring: str) -> Tuple[str, bool]:
+    """
+    Detect wildcard characters, and escape any that occur within a literal.
+
+    Parameters
+    ----------
+    querystring : str
+
+    Returns
+    -------
+    str
+        Query string with wildcard characters enclosed in literals escaped.
+    bool
+        If a non-literal wildcard character is present, returns True.
+
+    """
+    # This should get caught by the controller (form validation), but just
+    # in case we should check for it here.
+    if querystring.startswith('?') or querystring.startswith('*'):
+        raise QueryError('Query cannot start with a wildcard')
+
+    # Escape wildcard characters within string literals.
+    # re.sub() can't handle the complexity, sadly...
+    parts = re.split(STRING_LITERAL, querystring)
+    parts = [part.replace('*', r'\*').replace('?', r'\?')
+             if part.startswith('"') or part.startswith("'") else part
+             for part in parts]
+    querystring = "".join(parts)
+
+    # Only unescaped wildcard characters should remain.
+    wildcard = re.search(r'(?<!\\)([\*\?])', querystring) is not None
+    return querystring, wildcard
+
+
+def _Q(qtype: str, field: str, value: str) -> Q:
+    """Construct a :class:`.Q`, but handle wildcards first."""
+    value, wildcard = _wildcardEscape(value)
+    if wildcard:
+        return Q('wildcard', **{field: value})
+    return Q(qtype, **{field: value})
+
+
+class SearchSession(object):
+    """Encapsulates session with Elasticsearch host."""
+
+    # TODO: we need to take on security considerations here. Presumably we will
```

Comment:
Doesn't seem like we'd need HTTP auth for ElsasticSearch since it should be in a VPC (no one should be able to get to it) and all the data in it is public ( so worst case is like a DOS of a secondary arXiv function).

Reply:
Great thought. HTTPS was default for the AWS ES service (is HTTP an option?), so we went ahead with that. If we decide to move to our own cluster, this will be good to revisit. But yes, the ES endpoint is in a private VPC.